src/featbin/compute-filter.cc

```diff
@@ -0,0 +1,223 @@
+// featbin/compute-filters.cc
+
+// Copyright 2016 Pegah Ghahremani
+
+
+#include "feat/feature-functions.h"
+#include "matrix/srfft.h"
+#include "matrix/kaldi-matrix-inl.h"    
+#include "feat/wave-reader.h"
+namespace kaldi {
+
+// This function computes and accumulates  
+// dot-product of zero-mean input with its shifted version.
+// inner_prod[j] - inner product of input vector with its shifted version by j
+//                 sum_{i=0}^corr_window_size x[i] * x[i+j]
+// norms[j]      - inner product of shifted input by itself as sum_{i=0}^corr_window_size x[i+j]^2
+// lpc_order is the size of autocorr_coeffs.
+// If mean == 0, the wave is normalized using its mean, otherwise
+// it is mean-normalized using mean value.
+void AccStatsForCorrelation(const VectorBase<BaseFloat> &wave,
+                                int32 lpc_order, int32 corr_window_size,
+                                Vector<BaseFloat> *inner_prod,
+                                Vector<BaseFloat> *norms,
+                                BaseFloat mean) {
+  if (inner_prod->Dim() != lpc_order)
+    inner_prod->Resize(lpc_order);
+
+  if (norms->Dim() != lpc_order) 
+    norms->Resize(lpc_order);
+
+  Vector<BaseFloat> zero_mean_wave(wave);
+  if (mean != 0.0) 
+    zero_mean_wave.Add(mean);
+   else 
+    // subtract mean to normalize wave.
+    zero_mean_wave.Add(-wave.Sum() / wave.Dim());
+
+  // append zero et end if the length of shifted version 
+  // is less that corr_window_size.
+  zero_mean_wave.Resize(lpc_order + corr_window_size, kCopyData);
+  BaseFloat e1, e2, sum;
+  SubVector<BaseFloat> sub_vec1(zero_mean_wave, 0, corr_window_size);
+  e1 = VecVec(sub_vec1, sub_vec1);
+  (*norms)(0) += e1;
+  (*inner_prod)(0) += e1;
+  for (int32 lag = 1; lag < lpc_order; lag++) {
+    SubVector<BaseFloat> sub_vec2(zero_mean_wave, lag, corr_window_size);
```

Comment:
this way of computing e2 is very inefficient, as it only differs slightly from the previous e2.  better to just add the single-sample corrective terms.